python/mleap/pyspark/feature/math_unary.py

```diff
@@ -0,0 +1,84 @@
+from enum import Enum
+
+from pyspark import keyword_only
+from pyspark.ml.param.shared import HasInputCol, HasOutputCol
+from pyspark.ml.util import JavaMLReadable
+from pyspark.ml.util import JavaMLWritable
+from pyspark.ml.util import _jvm
+from pyspark.ml.wrapper import JavaTransformer
+
+from mleap.pyspark.py2scala import jvm_scala_object
+
+
+class UnaryOperation(Enum):
+    Sin = 1
+    Cos = 2
+    Tan = 3
+    Log = 4
+    Exp = 5
+    Abs = 6
+    Sqrt = 7
+
+
+class MathUnary(JavaTransformer, HasInputCol, HasOutputCol, JavaMLReadable, JavaMLWritable):
+
+    @keyword_only
+    def __init__(self, operation=None, inputCol=None, outputCol=None):
+        """
+        Computes the mathematical unary `operation` over the input column.
+
+        NOTE: we can't make `operation` a JavaParam (as in pyspark) because the
+            underlying scala object MathUnary uses a MathUnaryModel to store
+            the info about the unary operation (sin, tan, etc.)
+
+            If operation is a JavaParam, py4j will fail trying to set it on the
+            underlying scala object.
+
+            If operation doesn't have a default value, then pyspark will fail
+            upon deserialization trying to instantiate this object without args:
+                (it just runs py_type() where py_type is the class name)
+
+        """
+        super(MathUnary, self).__init__()
+
+        # if operation=None, it means that pyspark is reloading the model
+        # from disk and calling this method without args. In such case we don't
+        # need to set _java_obj here because pyspark will set it after creation
+        #
+        # if operation is not None, we can proceed to instantiate the scala classes
+        if operation:
+            scalaUnaryOperation = jvm_scala_object(
+                _jvm().ml.combust.mleap.core.feature.UnaryOperation,
+                operation.name
+            )
+
+            scalaMathUnaryModel = _jvm().ml.combust.mleap.core.feature.MathUnaryModel(scalaUnaryOperation)
+
+            self._java_obj = self._new_java_obj(
+                "org.apache.spark.ml.mleap.feature.MathUnary",
+                self.uid,
+                scalaMathUnaryModel,
+            )
+
+        self._setDefault()
```

Comment:
for my knowledge only, what does this do?

Reply:
So, the entire structure of this `JavaTransformer` is taken from the apache/spark project. Check their huge file of pyspark `features` here (Normalizer, NGram etc.):
https://github.com/apache/spark/blob/4a64901ab7176ad3d41c4598dea1e73cd92e969e/python/pyspark/ml/feature.py#L1827

As you can see I'm inheriting from JavaTransformer and implementing methods, calling `_setDefault()` like they are doing (I think it's a no-op but if the mother-project is calling it, I wanted to be consistent).

Too bad Spark doesn't provide MathUnary pyspark wrappers already.. This is also the reason for the [controversial](https://github.com/combust/mleap/pull/666/files#r414778794):
```python
sys.modules['pyspark.ml.mleap'] = mleap
```

Unfortunately [this code](https://github.com/apache/spark/blob/3ce1dff7ba4799295a2bf89e6df35e900afc587b/python/pyspark/ml/wrapper.py#L244) in Apache/spark
 searches for the python classes under the `pyspark` module. They don't let us inject the location of our wrapper's module from the outside, so we need to monkeypatch the code to make the MathUnary appear as if it was in the apache/spark project. It's similar to what MLeap does with the reference.conf file but less flexible 

One solution to this could be to subclass the JavaParams I linked above and override the entire `_from_java()` method, but I'm not sure how everyone feels about that..